### ISSUE TITLE
Question: Difficulties when creating the symmetries using SymSeg

### ISSUE THREAD

**Issue by Cznielsen** [chensong1995/HybridPose#13]
Hello,

First of all, thank you for your great work. We are writing to you and not the original author of SymSeg since you have recently used the library successfully.

We are trying to generate ground truth labels for synthetic data but have run into some issues when using SymSeg to generate the symmetry correspondences. In particular, objects which are clearly symmetric across some planes do not have any symmetries detected, perhaps due to the downsampling of the cloud?

Did you have to modify or tweak any of the code/config parameters before computing the symmetries to make it work?

Furthermore, we also get segmentation faults (bad alloc) when trying to run any of the actual symmetry segmentation scripts (only symmetry detection works). Do you recall which versions of the dependencies of SymSeg you used?

Thank you

**Comment by chensong1995** [comment]
Hello Cznielsen,

Thanks for your interest in our work!

1. From my experience, SymSeg requires very careful object-by-object hyperparameter tuning. The hyperparameters can be configured by editing `examples/reflectional_detection/config.pbtxt`. You will likely have to refer to their paper and/or C++ implementation for the meaning of the hyperparameters. Please also note that the input should contain a normal for each point in the point cloud.

2. I did not use any scripts in SymSeg other than the one for reflection symmetry detection. So sadly, I can't really help on their symmetry segmentation script. It might be a good idea to directly contact the authors of SymSeg for suggestions.

I hope this helps.

**Comment by Cznielsen** [comment]
Hello, Thank you for your response.

Forgive my ignorance, but if you only utilised the reflectional detection (to generate the symmetry plane), how did you then get the symmetry correspondences? By calculating the symmetric point of every other point across the plane manually?

**Comment by chensong1995** [comment]
> Hello, Thank you for your response.
> 
> It appears the missing symmetries stems from the lack of normals in our point clouds.
> Forgive my ignorance, but if you only utilised the reflectional detection (to generate the symmetry plane), how did you then get the symmetry correspondences? By calculating the symmetric point of every other point across the plane manually?

Yes, we project every point in the point cloud across the plane to find its symmetry correspondence in 3D. We then project 3D correspondences using the ground-truth pose to obtain 2D correspondences.

I hope this helps.

**Comment by Cznielsen** [comment]
Hello again,

We have managed to get the 2D coordinates (in u,v) of our 3D models (both original and the symmetric coordinates) by multiplying the transformation matrix described in the OcclusionDataset readme with the 3D local object coordinates of our model. Now we are having a bit of trouble generating the correspondences ourselves. In particular we are currently trying to replicate the structure of the correspondences you have generated.

We understand that the array is of the shape **[resolution_height, resolution_width, 2]**.
Our initial assumption was that the value of a given entry was the relative pixel coordinates/path to its symmetric counterpart in 2D camera pixel-coordinates, however after further investigation this does not appear to be the case. 

Can you explain to us the structure of your correspondence file?

To explain our thoughts regarding the above: if e.g. we load the first correspondence file of ape (0.npy) into some numpy array variable **A,** then the first value is located on 
**A[155, 414] = [-1, 2]**.
 Assumming -1 means down one pixel, then the next value is 
**A[156, 416] = [2, 1]**. 
However if we go one step further it will take us to 
**A[154, 417] = [0, 0]**
and thus a dead end.

**Comment by chensong1995** [comment]
Hello Cznielsen,

Thanks for your question!

`A[155, 414] = [-1, 2]` means there is a symmetry correspondence pair between (155, 414) and (155-1, 414+2). In other words, a symmetry correspondence pair between (155, 414) and (154, 417).

As you have observed, `A[154, 417] = [0, 0]`. Two possibilities:
1. Pixel (154, 417) is not part of the object. It is a background pixel.
2. There is a symmetry correspondence pair between (154, 417) and itself.

Projecting 3D symmetry correspondences to 2D is a little bit tricky. There are a few caveats:
1. Multiple 3D points with different depths will map to the same 2D point. In our released data, we only keep the one closest to the camera. 
2. The 2D projection of even a very dense 3D point cloud will leave some small holes on the image plane. This means if we want every single object pixel on the segmentation mask to have a correspondence, we will have to perform some interpolations either in 3D or 2D. In our experiments, we performed 2D interpolations to fill the holes.
3. When the object is only partially symmetric, the correspondence of an object pixel may be a background pixel. Imagine a human body whose left arm is pointing up and right arm is pointing down. There will be a symmetry correspondence pair between the left finger top and some pixel in the air to the right of the body.  

All the factors above suggest that a symmetry correspondence pair from pixel X to Y does not necessarily mean an identical correspondence pair from Y to X. They are not very hard to handle, but the code is pretty nasty. This is the biggest reason why I chose to not release our data labeling script.

I hope this helps.

**Comment by Cznielsen** [comment]
Thank you for your helpful response.

May I ask how you performed the interpolation in 2D?

I tried overlaying the mask with the projected points for both the original and the reflected model and using the remaining pixels, but as the projections are not completely alike (only very close), this does not yield any correspondences as far as I can tell.

The other option that I can think of would be to do the interpolation directly on the correspondences from the projections, but I am having trouble figuring out how to "connect" the newly interpolated pixels with their symmetric counterpart. The question is what the values/path of these interpolated pixels will be: Interpolated from their nearest neighbours or something else I haven't thought of?

Once again thank you for all your assistance.

**Comment by Cznielsen** [comment]
Thanks a lot it appears to be working more or less. 

As `inpaint` only takes 8 bit unsigned we simply added 128 to the values before converting them to the unsigned variant. After the inpainting we changed the type back to int16 before subtracting 128 again.

The interpolated values appear to be somewhat decent (with a number of inconsistencies). Hopefully this won't make a large difference as it is only a small number of the pixels that are not covered by the projected points. Our values were most similar when using the INPAINT_NS with a radius of 1 or 2, but we have not done a lot of testing regarding this.

Once again thank you for all your help.

**Comment by chensong1995** [comment]
> As `inpaint` only takes 8 bit unsigned ...

I believe `inpaint` takes more data types. Please take a look at [this](https://github.com/opencv/opencv/blob/ddcf3882704942ee79f5f5ff06e2eec678ee8262/modules/photo/src/inpaint.cpp#L808).

**Comment by Cznielsen** [comment]
> 
> 
> > As `inpaint` only takes 8 bit unsigned ...
> 
> I believe `inpaint` takes more data types. Please take a look at [this](https://github.com/opencv/opencv/blob/ddcf3882704942ee79f5f5ff06e2eec678ee8262/modules/photo/src/inpaint.cpp#L808).

We had already looked at the source code earlier. As far as we could tell the only data type which supported the two channels of the correspondence array was the 8 bit unsigned integers with 3 channels defined at line [753](https://github.com/opencv/opencv/blob/ddcf3882704942ee79f5f5ff06e2eec678ee8262/modules/photo/src/inpaint.cpp#L753)

**Comment by EasonEdison** [comment]
> **The explanation below has been edited because my original response has errors.**
> 
> Hello Cznielsen,
> 
> Thanks for your question!
> 
> `A[155, 414] = [-1, 2]` means there is a symmetry correspondence pair between (155, 414) and (155+2, 414-1). In other words, a symmetry correspondence pair between (155, 414) and (157, 413). Please note the ordering of coordinates. I was following the opencv convention, which is different from the natural ordering of numpy arrays.
> 
> Now, `A[157, 413] = [2, -2]`. This means there is a symmetry correspondence pair between (157, 413) and (157-2, 413+2). In other words, (157, 413) and (155, 415).
> 
> Projecting 3D symmetry correspondences to 2D is a little bit tricky. There are a few caveats:
> 
> 1. Multiple 3D points with different depths will map to the same 2D point. In our released data, we only keep the one closest to the camera.
> 2. The 2D projection of even a very dense 3D point cloud will leave some small holes on the image plane. This means if we want every single object pixel on the segmentation mask to have a correspondence, we will have to perform some interpolations either in 3D or 2D. In our experiments, we performed 2D interpolations to fill the holes.
> 3. When the object is only partially symmetric, the correspondence of an object pixel may be a background pixel. Imagine a human body whose left arm is pointing up and right arm is pointing down. There will be a symmetry correspondence pair between the left finger top and some pixel in the air to the right of the body.
> 4. There will be some numerical errors during the calculation.
> 
> All the factors above suggest that a symmetry correspondence pair from pixel X to Y does not necessarily mean an identical correspondence pair from Y to X. They are not very hard to handle, but the code is pretty nasty. This is the biggest reason why I chose to not release our data labeling script.
> 
> I hope this helps.

Thank you for your detailed explanation, and I have some questions:
In case 3, "There will be a symmetry correspondence pair between the left finger top and some pixel in the air to the right of the body.",  adopt such not symmetry correspondence or abandon it? 
And, label with  **correspondence pair between the left finger top and some pixel in the air to the right of the body** and **correspondence pair between the right finger down and some pixel in the air to the left of the body** is same?

**Comment by Cznielsen** [comment]
Hello again,

We had preliminary training but it appears some of our symmetries are bad. When creating symmetries files there were some inconsistencies. The primary inconsistency is that the x and y values of the symmetry file appear to be swapped. For example for duck 70:

Your symmetry (from occlusion linemod):
```
point:
   -6.87438e-05  -0.0129754  0.00240565  
normal:
   0.999924  0.00318167  -0.0118904  
```
Our symmetry file:
```
point:
   0.190385  -0.261637  -3.51017  
normal:
   0.00917105  0.999923  0.00835336  
```

When visualising the similarity is pretty obvious: 
[Your symmetry](https://cdn.discordapp.com/attachments/549261492327809024/695181816235032596/unknown.png) 
[Our symmetry](https://cdn.discordapp.com/attachments/549261492327809024/695181692511453314/unknown.pngl)
[Our symmetry with x and y swapped](https://cdn.discordapp.com/attachments/549261492327809024/695182066014355466/unknown.png).

Did you manually swap x and y coordinates of the point and normal vector in the symmetries file, or did it happen automatically somewhere (or maybe it happened for us somewhere)? Furthermore, it is possible for you to publish the symseg config file for reflectional detection, so that we may try it with your config?

Our issue is now that when visualizing the symmetry correspondences for our own models, flipping the symmetries is not enough, so we are looking into the fact that some of the axes of our may be flipped. Example:
[YCB Banana](https://cdn.discordapp.com/attachments/549261492327809024/694890343342276658/0_0_sym_gt.jpg)
[YCB Banana with x and y flipped in symmetries.txt](https://cdn.discordapp.com/attachments/549261492327809024/694890485646360606/0_0_sym_gt.jpg)

Thank you.

**Comment by chensong1995** [comment]
> In case 3, "There will be a symmetry correspondence pair between the left finger top and some pixel in the air to the right of the body.", adopt such not symmetry correspondence or abandon it?

We do not discard any symmetry correspondence during the labeling process. 

> And, label with **correspondence pair between the left finger top and some pixel in the air to the right of the body** and **correspondence pair between the right finger down and some pixel in the air to the left of the body** is same?

I am not exactly sure what you mean here. Can you please rephrase your question?

**Comment by chensong1995** [comment]
> Hello again,
> 
> We have done some preliminary training and it appears some of our symmetries are bad. When creating symmetries files there were some inconsistencies. The primary inconsistency is that the x and y values of the symmetry file appear to be swapped. For example for duck 70:
> 
> Your symmetry (from occlusion linemod):
> 
> ```
> point:
>    -6.87438e-05  -0.0129754  0.00240565  
> normal:
>    0.999924  0.00318167  -0.0118904  
> ```
> 
> Our symmetry file:
> 
> ```
> point:
>    0.190385  -0.261637  -3.51017  
> normal:
>    0.00917105  0.999923  0.00835336  
> ```
> 
> When visualising the similarity is pretty obvious:
> [Your symmetry](https://cdn.discordapp.com/attachments/549261492327809024/695181816235032596/unknown.png)
> [Our symmetry](https://cdn.discordapp.com/attachments/549261492327809024/695181692511453314/unknown.png)
> [Our symmetry with x and y swapped](https://cdn.discordapp.com/attachments/549261492327809024/695182066014355466/unknown.png).
> 
> Did you manually swap x and y coordinates of the point and normal vector in the symmetries file, or did it happen automatically somewhere (or maybe it happened for us somewhere)? Furthermore, it is possible for you to publish the symseg config file for reflectional detection, so that we may try it with your config?
> 
> Our issue is now that when visualizing the symmetry correspondences for our own models, flipping the symmetries is not enough, so we are looking into the fact that some of the axes of our may be flipped. Example:
> [YCB Banana](https://cdn.discordapp.com/attachments/549261492327809024/694890343342276658/0_0_sym_gt.jpg)
> [YCB Banana with x and y flipped in symmetries.txt](https://cdn.discordapp.com/attachments/549261492327809024/694890485646360606/0_0_sym_gt.jpg)
> 
> Thank you.

No, I did not flip the axes in the symmetries.txt file. Take a look at [this Python script](https://github.com/chensong1995/HybridPose/files/4424038/label.zip), which is used by me to generate custom labels on Occlusion Linemod.

I regret to say that the SymSeg configuration files we used to label the symmetry planes are lost at this moment.

**Comment by EasonEdison** [comment]
> > In case 3, "There will be a symmetry correspondence pair between the left finger top and some pixel in the air to the right of the body.", adopt such not symmetry correspondence or abandon it?
> 
> We do not discard any symmetry correspondence during the labeling process.
> 
> > And, label with **correspondence pair between the left finger top and some pixel in the air to the right of the body** and **correspondence pair between the right finger down and some pixel in the air to the left of the body** is same?
> 
> I am not exactly sure what you mean here. Can you please rephrase your question?

Thank you for your reply ,and my english is bad so can't explain clearly, so I use chinese this time.
我一开始以为只是从一边到另一边的对称映射，比如从右半身往左半身映射，所以就在想是从右向左映射还是从左向右映射，看了您的答复后，想到我应该是理解错了，应该是即从左边映射也从右边映射，在此感谢您的答复！

**Comment by chensong1995** [comment]
> Thank you for your reply ,and my english is bad so can't explain clearly, so I use chinese this time.
> 我一开始以为只是从一边到另一边的对称映射，比如从右半身往左半身映射，所以就在想是从右向左映射还是从左向右映射，看了您的答复后，想到我应该是理解错了，应该是即从左边映射也从右边映射，在此感谢您的答复！

Although English is the dominant communication tool on Github and in the research community, I personally do not think the language barrier should impede the propagation of knowledge and ideas. This repository welcomes discussions in either English or (Mandarin) Chinese. I really hope I can provide help in other languages as well, but unfortunately, English and Chinese are the only languages I know how to use.

我们对数据集的标注中包括一个物体和背景的分割罩。这个分割罩可以告诉我们，在每张图片里都有哪些像素是属于前景物体的。每一个属于前景物体的像素都会有一个对称点的标注。

**Comment by EasonEdison** [comment]
> > Thank you for your reply ,and my english is bad so can't explain clearly, so I use chinese this time.
> > 我一开始以为只是从一边到另一边的对称映射，比如从右半身往左半身映射，所以就在想是从右向左映射还是从左向右映射，看了您的答复后，想到我应该是理解错了，应该是即从左边映射也从右边映射，在此感谢您的答复！
> 
> Although English is the dominant communication tool on Github and in the research community, I personally do not think the language barrier should impede the propagation of knowledge and ideas. This repository welcomes discussions in either English or (Mandarin) Chinese. I really hope I can provide help in other languages as well, but unfortunately, English and Chinese are the only languages I know how to use.
> 
> 我们对数据集的标注中包括一个物体和背景的分割罩。这个分割罩可以告诉我们，在每张图片里都有哪些像素是属于前景物体的。每一个属于前景物体的像素都会有一个对称点的标注。

感谢您的回答，我现在明白了，您之前在“we only keep the one closest to the camera.”里提到过，但是我后来又忘记了，这个的意思就是只选mask上的像素，而这些像素就是离摄像机最近的，对这些计算对称对应，是这样吗？

**Comment by chensong1995** [comment]
> 感谢您的回答，我现在明白了，您之前在“we only keep the one closest to the camera.”里提到过，但是我后来又忘记了，这个的意思就是只选mask上的像素，而这些像素就是离摄像机最近的，对这些计算对称对应，是这样吗？

是的。
